### DISCUSSION TITLE
r.learn.ml installation

### DISCUSSION THREAD

**Discussion by gdallenogare** [OSGeo/grass#2808] [Q&A] [answered]
It's my first time and I am very sorry for this basic question.
I am working on Debian, and GRASS GIS is installed by Package manager.
I don't know how to figure out and find a solution for this error.
Thanks to everyone can help me :) and thanks to this community
# =======================================
/usr/lib/grass78/scripts/g.extension extension=r.learn.ml operation=add         
WARNING: Extension <r.learn.ml> already installed. Re-installing...
Fetching <r.learn.ml> from GRASS GIS Addons repository (be patient)...
svn: E170000: URL 'https://github.com/OSGeo/grass-
addons/trunk/grass7/raster/r.learn.ml' doesn't exist
ERROR: GRASS Addons <r.learn.ml> not found

**Comment by nilason** [comment]
Looks like an old version of GRASS.

What version is it?

(`grass --version` or from a grass session `g.version`)

**Reply by gdallenogare** [reply]
GRASS 7.8.5 (epsg25830):~ > g.version
GRASS 7.8.5 (2020)
[Raster MASK present]

From the Debian stable repository

**Reply by gdallenogare** [reply]
I found this way to install it:
g.extension extension=r.learn.ml operation=add url=https://svn.osgeo.org/grass/grass-addons/grass7/
But when I run from the GUI, in the command output is printed: "ERROR: Scikit learn 0.18 or newer is not installed" ... even it's already installed. I can call scikit from the python console without errors (scikit-learn 1.2.1).

**Reply by nilason** [reply]
There has since (if I remember correct) 7.8.5 been infrastructural changes with GRASS addons, so the URL is no more valid.
It should work with 7.8.7 (and maybe even 7.8.6) or even better the latest 8.2.1. I don't know how readily they are available on Debian though.

**Reply by nilason** [reply]
The addons found on the url  https://svn.osgeo.org/grass/grass-addons/grass7/ are a bit dated compared to latest. But should work.

You could try within GRASS:

`$GRASS_PYTHON -m pip install scikit-learn`

**Reply by neteler** [reply]
> There has since (if I remember correct) 7.8.5 been infrastructural changes with GRASS addons, so the URL is no more valid. It should work with 7.8.7 (and maybe even 7.8.6) or even better the latest 8.2.1. I don't know how readily they are available on Debian though.

Here is the current list of versions: https://repology.org/project/grass/versions

**Reply by nilason** [reply]
> > There has since (if I remember correct) 7.8.5 been infrastructural changes with GRASS addons, so the URL is no more valid. It should work with 7.8.7 (and maybe even 7.8.6) or even better the latest 8.2.1. I don't know how readily they are available on Debian though.
> 
> Here is the current list of versions: https://repology.org/project/grass/versions

Yes, I did look at that one. Debian is seemingly very conservative with updates.

**Reply by nilason** [reply]
> But when I run from the GUI, in the command output is printed: "ERROR: Scikit learn 0.18 or newer is not installed" ... even it's already installed. I can call scikit from the python console without errors (scikit-learn 1.2.1).

This error is also caused by a too new version of scikit-learn, regardless the content of the message.

**Reply by gdallenogare** [reply]
> The addons found on the url https://svn.osgeo.org/grass/grass-addons/grass7/ are a bit dated compared to latest. But should work.
> 
> You could try within GRASS:
> 
> `$GRASS_PYTHON -m pip install scikit-learn`

It didn't work

**Reply by nilason** [reply]
> > You could try within GRASS:
> > `$GRASS_PYTHON -m pip install scikit-learn`
> 
> It didn't work

No, the installed `scikit-learn` is too new and thus not compatible with the addon version at https://svn.osgeo.org/grass/grass-addons/grass7/

**Comment by nilason** [comment]
I'm not expert on `g.extension`, but here's a cheap trick that might work:

replace the file: `$GRASS_ADDON_BASE/scripts/r.learn.ml` (note: there should be no .py file extension).

with: 
https://github.com/OSGeo/grass-addons/blob/grass7/src/raster/r.learn.ml/r.learn.ml.py

**Reply by gdallenogare** [reply]
It didn't work

**Comment by gdallenogare** [comment, marked as answer]
My drastic/ignorant solution: updating Debian from stable to testing :) 
https://itsfoss.com/switch-debian-stable-testing/
So far, everything is working pretty well ... I hope, it will be in future as well

Thanks Nicklas and Markus for your precious help

**Reply by nilason** [reply]
Good!
If upgrading Debian wasn't an option, you could have compiled a newer version of GRASS following instructions on https://grasswiki.osgeo.org/wiki/Compile_and_Install#Debian

**Reply by gdallenogare** [reply]
I see and thanks to remember me to this option. I switched to tasting release also for other conflicts between qgis and python that probably are fixed (already tested within a container https://github.com/89luca89/distrobox ) .. crossing fingers

**Reply by gdallenogare** [reply]
Update: In the end I've rolled back to the stable release and installed the compilated version :)
